### PR TITLE
Have generic SENTs check the gamemode SENT limit.

### PR DESCRIPTION
**Pull request by thegrb93**
Otherwise infinite bouncy balls can be pasted.

### CODE REVIEW

File: lua/advdupe2/sv_clipboard.lua [hunk 1]
```diff
@@ -912,6 +912,7 @@ local function CreateEntityFromTable(EntTable, Player)
 				valid:RestoreNetworkVars(EntTable.DT)
 			end
 
+			Player:AddCount("sents", valid)
```
Comment (AbigailBuccaneer) [inline lua/advdupe2/sv_clipboard.lua]: This should be handled by calling the `GM:PlayerSpawnedSENT` hook instead.

**Comment by thegrb93** [comment]
Tested it much more thoroughly. Should work much better now.

File: lua/advdupe2/sv_clipboard.lua [hunk 2]
```diff
@@ -911,16 +908,15 @@ local function CreateEntityFromTable(EntTable, Player)
...
-						gamemode.Call("PlayerSpawnedEffect", Player, valid:GetModel(), valid)
-					else
-						gamemode.Call("PlayerSpawnedSENT", Player, valid)
-					end
+			
+			if GENERIC then
+				if(EntTable.Class=="prop_effect")then
+					gamemode.Call("PlayerSpawnedEffect", Player, valid:GetModel(), valid)
+				else
+					gamemode.Call("PlayerSpawnedSENT", Player, valid)
```
Comment (AbigailBuccaneer) [inline lua/advdupe2/sv_clipboard.lua]: The full list is in [gamemodes/sandbox/gamemode/player.lua](https://github.com/garrynewman/garrysmod/blob/3e138636eb1b0ad6ed785dedf350020755cff5f1/garrysmod/gamemodes/sandbox/gamemode/player.lua):

* `GM:PlayerSpawnedRagdoll`
* `GM:PlayerSpawnedProp`
* `GM:PlayerSpawnedEffect`
* `GM:PlayerSpawnedVehicle`
* `GM:PlayerSpawnedNPC`
* `GM:PlayerSpawnedSENT`
* `GM:PlayerSpawnedSWEP`

Unless there's a reason I'm missing that we don't have to, we should handle all of these.

Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua]: status,valid = pcall(  EntityClass.Func, Player, unpack(ArgList) )

Handles all other cases.

**Comment by thegrb93** [comment]
No issues.

**Comment by AbigailBuccaneer** [comment]
Thanks for your work, @thegrb93 !

**Comment by Mista-Tea** [comment]
Is calling PlayerSpawnSENT for literally any entity that isn't a vehicle/npc/ragdoll really intended behavior? Shouldn't PlayerSpawnSENT only be called for things that are actual SENTs, i.e. things that are registered as "sents" in cleanup.Add()/PLAYER:AddCount() and everything inside the "Entities" tab?

I ask because this PR (lines https://github.com/wiremod/advdupe2/pull/70/files#diff-c7eb62437d84a1be1594e19073cb7752R866 and 867) has entirely broken the AdvDupe2 for me because I have hooks into PlayerSpawnSENT to block players from spawning actual SENTs (sent_ball, ammo, balloondeployer, etc).

Now when players spawn anything from gmod_* (including gmod_wire_*), PlayerSpawnSENT is getting called when it shouldn't be and is causing everything to be blocked. Gmod_wheels, gmod_thrusters and the like aren't "sents", they are "wheels", "thrusters", etc.

**Comment by Mista-Tea** [comment]
The problem also stems from a prior PR - https://github.com/thegrb93/advdupe2/commit/61e1f372e020717e58de61ca6e066f8102986fb7.

PlayerSpawnEntity was meant to be a custom hook, not PlayerSpawnSENT. I relied on it so I could use my own logic and call PLAYER:CheckLimit() with the appropriate type ("sents", "npcs", "sweps", etc) for my limit checking. 

It was only called for entities spawned through AdvDupe2, leaving hooks like PlayerSpawnSENT to be properly called for actual SENTs. You don't see PlayerSpawnSENT being called when spawning gmod_* stuff (lights/thrusters/wheels/etc) from within the Q menu, so why call it here?

**Comment by thegrb93** [comment]
Yes I know PlayerSpawnEntity was a custom hook, but restriction mods didn't use it and players were allowed to spawn entities that were restricted using advdupe2. I think if you adapted your mods to use PlayerSpawnEntity then you can readapt them for PlayerSpawnSENT. And gmod_*, gmod_wire_* entities technically are SENTS so I don't see the problem.

**Comment by thegrb93** [comment]
The only "proper" way I can think of doing it is to figure out which tool the entity is spawned from and use the "CanTool" hook, but I don't think that is possible.

**Pull request merged**